fhir-codegen/gradle-plugin/src/main/kotlin/com/google/fhir/codegen/EnumTypeSpecGenerator.kt

```diff
@@ -0,0 +1,288 @@
+            FunSpec.builder("getDisplay")
+              .addModifiers(KModifier.PUBLIC)
+              .returns(String::class.asClassName().copy(nullable = true))
+              .addStatement("return display")
+              .build()
+          )
+          addFunction(
+            FunSpec.builder("getDefinition")
+              .addModifiers(KModifier.PUBLIC)
+              .returns(String::class.asClassName().copy(nullable = true))
+              .addStatement("return definition")
+              .build()
+          )
+          addType(
+            TypeSpec.companionObjectBuilder()
+              .addFunction(
+                FunSpec.builder("fromCode")
+                  .addParameter("code", String::class)
+                  .returns(ClassName.bestGuess(enumClassName))
+                  .beginControlFlow("return when (code)")
+                  .apply {
+                    fhirEnum.constants.forEach { addStatement("%S -> %L", it.code, it.name) }
+                    addStatement(
+                      "else -> throw IllegalArgumentException(\"Unknown code \$code for enum %L\")",
+                      enumClassName,
+                    )
+                  }
+                  .endControlFlow()
+                  .build()
+              )
+              .build()
+          )
+        }
+        .build()
+    return typeSpec
+  }
+
+  /**
+   * Instantiate a [FhirEnum] to facilitate the generation of Kotlin enum classes. The enum
+   * constants are derived from concepts defined in a `ValueSet`, a `CodeSystem`, or both. When both
+   * are provided, the constants are generated by intersecting the concepts from the `ValueSet` and
+   * the `CodeSystem`. If only one is provided, the constants are based solely on that source. Any
+   * nested concepts are flattened, transformed, and included in the resulting enum constants.
+   *
+   * NOTE: This excludes all systems that equals "http://unitsofmeasure.org" or start with "urn",
+   * e.g. urn:ietf:bcp:13, urn:ietf:bcp:47,urn:iso:std:iso:4217 typically used for MIMETypes,
+   * Currency Code etc.
+   */
+  private fun generateEnum(valueSet: ValueSet, codeSystemMap: Map<String, CodeSystem>): FhirEnum? {
+    return valueSet.compose
+      ?.include
+      ?.filter { it.isValueSystemSupported() }
+      ?.let {
+        val enumConstants = mutableListOf<FhirEnumConstant>()
+        for (include in it) {
+          val system = include.system!!
+          val codeSystem = codeSystemMap[system]
+
+          generateEnumConstants(
+            system = system,
+            codeSystemConcepts = codeSystem?.concept,
+            valueSetConcepts = include.concept,
+            enumConstants = enumConstants,
+          )
+        }
+        return FhirEnum(valueSet.description, enumConstants)
+      }
+  }
+
+  private fun generateEnumConstants(
+    system: String,
+    codeSystemConcepts: List<Concept>?,
+    valueSetConcepts: List<Concept>?,
+    enumConstants: MutableList<FhirEnumConstant>,
```

Comment:
i don't love this idea of carrying this mutable list and keeping appending to it. can you just make this function return a list? and then you can use flatMap in the function above. in general resulting in more idiomatic and "functional" (as in functional programming) code.